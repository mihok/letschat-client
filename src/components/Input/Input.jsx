import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import './Input.css';

import { sendMessage } from '../../containers/Chat/actions';
import { CHAT_DISCONNECTED } from '../../containers/Chat/constants';


// TODO: move to a constants file?
const KEY_ENTER = 13;

export class InputComponent extends Component {
  static propTypes = {
    chatStyle: PropTypes.string,
    chatStatus: PropTypes.string,
    chatId: PropTypes.string,
    clientId: PropTypes.string,
    dispatch: PropTypes.func,
    socket: (props, propName) => {
      if (!(propName in props)) {
        throw new Error('socket must be set.');
      }
    },
  }


  constructor (props) {
    super(props);
    this.socket = props.socket;

    this.state = { messageBox: '' };
  }


  onKeyPress = (event) => {
    const { key, keyCode, shiftKey, ctrlKey, altKey } = event;
    const { chatId, clientId } = this.props;
    console.log(`INPUT KEYPRESS ${key} (${keyCode}), SHIFT ${shiftKey}, CTRL ${ctrlKey}, ALT ${altKey}`);

    if (keyCode === KEY_ENTER) {
      if (!shiftKey) {
        console.log('SENDING MESSAGE ...');
        // Send data
        this.socket.emit('client:message', JSON.stringify({
          author: `client.${clientId}`,
          content: event.target.value,
          chat: chatId,
          timestamp: (new Date()).toISOString(),
        }));
        this.props.dispatch(sendMessage(event.target.value));
        this.setState({ messageBox: '' });

        event.preventDefault();
      }
    }

    this.socket.emit('client:typing');
  }

  handleChange = (e) => { this.setState({ messageBox: e.target.value }); }

  render () {
    const { chatStyle, chatStatus } = this.props;
    const isDisabled = chatStatus === CHAT_DISCONNECTED;

    return (
      <textarea
        type="text"
        className={`input-${chatStyle.toLowerCase()}`}
        placeholder="Type a message&hellip;"
        onKeyDown={this.onKeyPress}
        onChange={this.handleChange}
        value={this.state.messageBox}
        disabled={isDisabled}
      />
    );
  }
}

const mapStateToProps = state => ({
  chatStyle: state.ui.style,
  chatStatus: state.chat.status,
  chatId: state.chat.session.id,
  clientId: state.chat.session.client.id,
});

const mapDispatchToProps = dispatch => ({
  dispatch,
});

const Input = connect(
  mapStateToProps,
  mapDispatchToProps,
)(InputComponent);

export default Input;
