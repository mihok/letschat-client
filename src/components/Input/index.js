import { h } from 'preact';
import PropTypes from 'prop-types';

import { applyTheme } from '../ThemeProvider';

import './styles.css';

const Input = props => {
  const { sendMessage, theme, textBox, handleInput } = props;
  return (
    <form className={`Input__form--${theme}`} onSubmit={sendMessage}>
      <input
        className={`Input--${theme}`}
        placeholder="Type Here"
        onChange={e => handleInput(e)}
        name="messages"
        value={textBox}
      />
    </form>
  );
};

Input.propTypes = {
  handleInput: PropTypes.func,
  sendMessage: PropTypes.func,

  theme: PropTypes.string,
  textBox: PropTypes.string,
};

export default applyTheme(Input);
