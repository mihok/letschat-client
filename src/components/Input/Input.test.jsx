import React from 'react';

import { shallow, mount } from 'enzyme';

import Input from './Input.jsx';

const socket = {
  // on: jest.fn(),
  emit: jest.fn(),
};

const store = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
  getState: jest.fn(() => ({
    ui: { style: 'TEST' },
    chat: {
      status: 'TEST',
      session: {
        id: 'TEST123',
        client: { id: 'TEST123' },
      },
    },
  })),
};

describe('Input', () => {
  afterEach(() => {
    socket.emit.mockReset();
  });

  it('matches snapshot', () => {
    const component = shallow(<Input store={store} socket={socket} />);
    expect(component).toMatchSnapshot();
  });

  it('emits a client:typing event on keyDown', () => {
    const component = mount(<Input store={store} socket={socket} />);
    component.find('textarea').simulate('keyDown', {
      keyCode: 84,
      key: 't',
    });

    expect(socket.emit).toHaveBeenCalled();
    expect(socket.emit).toHaveBeenCalledWith('client:typing');
  });

  it('emits a client:message event on keyDown', () => {
    const component = mount(<Input store={store} socket={socket} />);

    component.find('textarea').simulate('keyDown', {
      keyCode: 13,
      shiftKey: false,
      key: 'Enter',
    });

    expect(socket.emit).toHaveBeenCalledTimes(2);
    expect(socket.emit.mock.calls[0][0]).toBe('client:message');
  });
});
