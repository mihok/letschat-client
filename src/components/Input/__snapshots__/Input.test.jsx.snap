exports[`Input matches snapshot 1`] = `
<InputComponent
  chatId="TEST123"
  chatStatus="TEST"
  chatStyle="TEST"
  clientId="TEST123"
  dispatch={[Function]}
  socket={
    Object {
      "emit": [Function],
    }
  }
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  } />
`;
